File: 2024/Dec 2024 Python/Day 5/day5.py
def main():

    fileName = "../textfiles/day5final.txt"

    part_one(fileName)


def part_one(fileName):
    d = dict()
    with open(fileName, encoding="utf8") as f:
        lines = f.read().splitlines()
        first_part = True
        cnt = 0
        for line in lines:
            if first_part and (line.isspace() or len(line) == 0):
                first_part = False
                continue
            if first_part:
                k = int(line.split("|")[0])
                v = int(line.split("|")[1])
                if k in d:
                    d[k].add(v)
                else:
                    d[k] = set()
                    d[k].add(v)
            else:
                nums = [int(x) for x in line.split(",")]
                # print(nums)
                stop = False
                numberOfMatches = 0
                for i, x in enumerate(nums):
                    if stop:
                        break
                    # print(i)
                    for _, y in enumerate(nums[i+1:]):
                        # print(x, y)
                        if x in d and y in d[x]:
                            numberOfMatches += 1
                        elif y in d and x in d[y]:
                            stop = True
                            break
                # print(numberOfMatches)
                if numberOfMatches == (len(nums)-1)*(len(nums))/2:
                    cnt += nums[int(len(nums)/2)]

        print("Day 5 part one", cnt)


main()

File: 2019/CPP Feb 2025/Day 5/day5.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>

#include "day5.h"

int main()
{
    std::string fileName{ "../textFiles/day5final.txt" };
    partOne(fileName);
    partTwo(fileName);
}

void partOne(std::string fileName)
{
    std::ifstream file{ fileName };

    if (file.is_open())
    {
        std::string line{};
        std::vector<int> codes{};
        std::string tok{};
        while (std::getline(file, line))
        {
            std::stringstream ss{ line };

            while (std::getline(ss, tok, ','))
            {
                codes.push_back(std::stoi(tok));
            }
        }
        int currentPos{ 0 };
        // make sure to read the whole problem statement
        codes[1] = 12;
        codes[2] = 2;

        while (codes[currentPos] != 99)
        {
            int a{ codes[codes[currentPos + 1]] };
            int b{ codes[codes[currentPos + 2]] };
            if (codes[currentPos] == 1)
            {
                codes[codes[currentPos + 3]] = a + b;
            }
            else if (codes[currentPos] == 2)
            {
                codes[codes[currentPos + 3]] = a * b;
            }
            currentPos += 4;
        }
        std::cout << "day 2 part one: " << codes[0] << std::endl;
    }
    else
    {
        std::cout << "file won't open" << std::endl;
    }
}

// void partTwo(std::string fileName)
// {
//     std::ifstream file{ fileName };

//     if (file.is_open())
//     {
//         std::string line{};
//         std::vector<int> codes{};
//         std::string tok{};
//         while (std::getline(file, line))
//         {
//             std::stringstream ss{ line };

//             while (std::getline(ss, tok, ','))
//             {
//                 codes.push_back(std::stoi(tok));
//             }
//         }

//         for (int noun = 0; noun < 100; noun++)
//         {
//             for (int verb = 0; verb < 100; verb++)
//             {

//                 std::vector newCodes = codes;
//                 // getAnswerForPartTwo(newCodes, 12, 12);
//                 if (getAnswerForPartTwo(newCodes, noun, verb))
//                 {
//                     std::cout << "Day 2 part two: " << noun * 100 + verb << std::endl;
//                     return;
//                 }
//             }
//         }
//         // std::cout << "noun: " << noun << std::endl;
//         // std::cout << "verb: " << verb << std::endl;
//         std::cout << "Day 2 part 2: " << "No answer found" << std::endl;
//     }
//     else
//     {
//         std::cout << "file won't open" << std::endl;
//     }
// }

// bool getAnswerForPartTwo(std::vector<int> codes, int noun, int verb)
// {
//     int currentPos{ 0 };
//     codes[1] = noun;
//     codes[2] = verb;
//     while (codes[currentPos] != 99)
//     {
//         int a{ codes[codes[currentPos + 1]] };
//         int b{ codes[codes[currentPos + 2]] };
//         if (codes[currentPos] == 1)
//         {
//             codes[codes[currentPos + 3]] = a + b;
//         }
//         else if (codes[currentPos] == 2)
//         {
//             codes[codes[currentPos + 3]] = a * b;
//         }
//         currentPos += 4;
//     }
//     return codes[0] == 19690720;
// }